docs/proposals/21-ipv6-singlestack-local.md

```diff
@@ -0,0 +1,213 @@
+kind: DNSRecord
+spec:
+  # ...
+  recordType: AAAA
+  values:
+  - dead::1
+```
+
+- `spec.recordType` allows specifying `AAAA` in addition to the current set of valid record types
+- `spec.values` allows specifying IPv6 values if `spec.recordType=AAAA`
+
+### Implementation Overview
+
+This section gives a rough overview of the most important changes for handling the `ipFamilies` fields.
+We won't go into detail here but only highlight a few selected changes.
+The most important takeaway is, that all components need to respect the `ipFamilies` fields.
+
+#### Preparing the Local Setup
+
+The kind cluster (used as garden runtime and seed cluster) is switched to IPv6 single-stack networking.
+This includes the kind cluster configuration itself as well as its calico configuration.
+Developers will need to perform some configuration of their machine and Docker installation.
+Corresponding instructions are made available for macOS and Linux machines.
+
+provider-local is configured to use IPv6 addresses for patching the status of LoadBalancer services.
+This is auto-detected if provider-local runs on the seed itself and configured explicitly if running on the host machine (which might have both IPv4 and IPv6 addresses).
+
+#### Scheduling
+
+gardener-apiserver only allows scheduling shoots to seeds with matching `ipFamilies` settings.
+I.e., all `Shoot.spec.networking.ipFamilies` values must be included in `Seed.spec.networks.ipFamilies`.
+This might be changed later on if "networking magic" is in place to host dual-stack shoots on IPv6 singe-stack seeds or similar.
+
+Also, gardener-scheduler chooses seeds with `ipFamilies` settings matching the shoot's configuration.
+
+#### DNS Records
+
+gardenlet creates `DNSRecord` objects for the shoot API server with the record type corresponding to the `Shoot.spec.networking.ipFamilies` setting (`A` for IPv4, `AAAA` for IPv6).
+`AAAA` records are created with a suffix to avoid name collisions in a potential dual-stack implementation.
+
+#### Network Policies
+
+Gardener manages several `NetworkPolicy` objects with static CIDRs (e.g. `allow-to-public-networks` in seeds).
+For these, IPv6-equivalents of the IPv4 CIDRs are added, e.g., `::/0` as an equivalent to `0.0.0.0/0`.
+
+#### Shoot Worker Node Kernel Configuration
+
+Kubernetes networking requires IPv6 forwarding to be enabled on the OS level.
+Hence, gardenlet explicitly enables the corresponding kernel setting for shoot worker nodes via `OperatingSystemConfigurations`, similar to IPv4 (ref [gardener/gardener#7046](https://github.com/gardener/gardener/pull/7046)).
+
+#### Docker Hub Images
+
+The `docker.io` registry doesn't support pulling images over IPv6 (see [Beta IPv6 Support on Docker Hub Registry](https://www.docker.com/blog/beta-ipv6-support-on-docker-hub-registry/)).
+
+Container images from `docker.io` used on shoots and seeds are rewritten to `registry.ipv6.docker.com` if the corresponding `ipFamilies` field specifies IPv6 singe-stack.
+
```

Comment:
This is only necessary if we don't have DNS64/464XLAT+NAT64 support which I would prefer over this hack.

Reply:
Yes, one could do some "network magic" as well. However, I don't consider the proposal a hack but rather the appropriate adaptation to the environment.
Also, the proposed solution is much simpler (as it doesn't involve any sophisticated setup on the host/docker VM) and is less error-prone.

Reply:
@philsbln i cant imagine how a developer could potentially make DNS64/464XLAT+NAT64 support work locally. An we talk about local-garden setup actually.

Reply:
Shipping a NAT64/DNS64 setup as a pair of root/host-networking containers is not that hard to do … probably a day of work for me given there is a ready-to-tun dev setup there ti start in the first place.